File: tools/cv-importer/src/portfolio_cv_importer/extractors/pdf_text_extractor.py
"""Pure Python extractor строк из PDF без OCR."""

from __future__ import annotations

import re
import zlib

from portfolio_cv_importer.normalizers.text_normalizer import normalize_extracted_line


def extract_lines_from_pdf_bytes(pdf_bytes: bytes) -> list[str]:
    """Извлекает человекочитаемые строки из PDF через ToUnicode-карты."""

    pdf_objects = parse_pdf_objects(pdf_bytes)
    font_to_unicode_maps = build_font_to_unicode_maps(pdf_objects)
    page_lines: list[str] = []

    for body_text in pdf_objects.values():
        if not re.search(r"/Type/Page\b", body_text):
            continue

        content_reference_match = re.search(r"/Contents\s+(\d+)\s+0\s+R", body_text)
        if content_reference_match is None:
            continue

        page_font_references = parse_page_font_references(body_text)
        content_object = pdf_objects.get(int(content_reference_match.group(1)))
        if content_object is None:
            continue

        content_stream = inflate_object_stream(content_object)
        if content_stream is None:
            continue

        text_items = extract_text_items(content_stream, page_font_references, font_to_unicode_maps)
        page_lines.extend(group_text_items_into_lines(text_items))

    normalized_lines = [normalize_extracted_line(line) for line in page_lines]
    return [line for line in normalized_lines if line]


def parse_pdf_objects(pdf_bytes: bytes) -> dict[int, str]:
    """Извлекает все объекты PDF как блоки latin1-текста."""

    pdf_text = pdf_bytes.decode("latin1")
    object_pattern = re.compile(r"(\d+)\s+(\d+)\s+obj")
    pdf_objects: dict[int, str] = {}

    for object_match in object_pattern.finditer(pdf_text):
        object_id = int(object_match.group(1))
        body_start = object_match.end()
        object_end = pdf_text.find("endobj", body_start)
        if object_end == -1:
            continue

        pdf_objects[object_id] = pdf_text[body_start:object_end]

    return pdf_objects


def inflate_object_stream(body_text: str) -> str | None:
    """Разжимает stream-часть PDF-объекта."""

    stream_index = body_text.find("stream")
    if stream_index == -1:
        return None

    stream_start_index = stream_index + len("stream")
    if body_text[stream_start_index:stream_start_index + 2] == "\r\n":
        stream_start_index += 2
    elif body_text[stream_start_index:stream_start_index + 1] == "\n":
        stream_start_index += 1

    stream_end_index = body_text.find("endstream", stream_start_index)
    if stream_end_index == -1:
        return None

    normalized_end_index = stream_end_index
    if body_text[normalized_end_index - 1:normalized_end_index] == "\n":
        normalized_end_index -= 1
    if body_text[normalized_end_index - 1:normalized_end_index] == "\r":
        normalized_end_index -= 1

    compressed_bytes = body_text[stream_start_index:normalized_end_index].encode("latin1")
    return zlib.decompress(compressed_bytes).decode("utf-8")


def build_font_to_unicode_maps(pdf_objects: dict[int, str]) -> dict[int, dict[str, str]]:
    """Строит карту фонтов и соответствующих ToUnicode-таблиц."""

    font_maps: dict[int, dict[str, str]] = {}

    for object_id, body_text in pdf_objects.items():
        to_unicode_match = re.search(r"/ToUnicode\s+(\d+)\s+0\s+R", body_text)
        if to_unicode_match is None:
            continue

        unicode_object = pdf_objects.get(int(to_unicode_match.group(1)))
        if unicode_object is None:
            continue

        unicode_stream = inflate_object_stream(unicode_object)
        if unicode_stream is None:
            continue

        font_maps[object_id] = parse_to_unicode_map(unicode_stream)

    return font_maps


def parse_to_unicode_map(unicode_stream: str) -> dict[str, str]:
    """Читает ToUnicode CMap и превращает его в словарь символов."""

    unicode_map: dict[str, str] = {}
    blocks = re.findall(r"(?:\d+)\s+begin(?:bfchar|bfrange)[\s\S]*?end(?:bfchar|bfrange)", unicode_stream)
    for block in blocks:
        block_lines = [line.strip() for line in block.splitlines() if line.strip()]
        for block_line in block_lines:
            char_match = re.match(r"^<([0-9A-F]+)>\s+<([0-9A-F]+)>$", block_line, flags=re.IGNORECASE)
            if char_match:
                unicode_map[char_match.group(1).upper()] = unicode_from_hex(char_match.group(2))
                continue

            range_match = re.match(
                r"^<([0-9A-F]+)>\s+<([0-9A-F]+)>\s+<([0-9A-F]+)>$",
                block_line,
                flags=re.IGNORECASE,
            )
            if range_match:
                range_start = int(range_match.group(1), 16)
                range_end = int(range_match.group(2), 16)
                target_code = int(range_match.group(3), 16)
                for source_code in range(range_start, range_end + 1):
                    unicode_map[source_code_to_key(source_code, len(range_match.group(1)))] = unicode_from_hex(
                        format(target_code, "x"),
                    )
                    target_code += 1
                continue

            array_match = re.match(
                r"^<([0-9A-F]+)>\s+<([0-9A-F]+)>\s+\[(.+)\]$",
                block_line,
                flags=re.IGNORECASE,
            )
            if array_match is None:
                continue

            range_start = int(array_match.group(1), 16)
            unicode_targets = re.findall(r"<([0-9A-F]+)>", array_match.group(3), flags=re.IGNORECASE)
            for target_index, unicode_target in enumerate(unicode_targets):
                unicode_map[source_code_to_key(range_start + target_index, len(array_match.group(1)))] = unicode_from_hex(
                    unicode_target,
                )

    return unicode_map


def parse_page_font_references(body_text: str) -> dict[str, int]:
    """Разбирает справочник font-ресурсов конкретной страницы."""

    font_references: dict[str, int] = {}
    font_block_match = re.search(r"/Font\s*<<([\s\S]*?)>>", body_text)
    if font_block_match is None:
        return font_references

    for font_match in re.finditer(r"/(F\d+)\s+(\d+)\s+0\s+R", font_block_match.group(1)):
        font_references[font_match.group(1)] = int(font_match.group(2))

    return font_references


def extract_text_items(
    content_stream: str,
    page_font_references: dict[str, int],
    font_to_unicode_maps: dict[int, dict[str, str]],
) -> list[dict[str, object]]:
    """Извлекает текстовые элементы с координатами для последующей сборки строк."""

    text_items: list[dict[str, object]] = []
    current_font_name: str | None = None
    current_text_x = 0.0
    current_text_y = 0.0

    for content_line in content_stream.splitlines():
        font_match = re.search(r"/(F\d+)\s+[\d.]+\s+Tf", content_line)
        if font_match is not None:
            current_font_name = font_match.group(1)

        transform_match = re.search(
            r"([\d.-]+)\s+[\d.-]+\s+[\d.-]+\s+[\d.-]+\s+([\d.-]+)\s+([\d.-]+)\s+Tm",
            content_line,
        )
        if transform_match is not None:
            current_text_x = float(transform_match.group(2))
            current_text_y = float(transform_match.group(3))

        unicode_map: dict[str, str] = {}
        if current_font_name is not None and current_font_name in page_font_references:
            unicode_map = font_to_unicode_maps.get(page_font_references[current_font_name], {})

        for array_match in re.finditer(r"\[(.*?)\]\s*TJ", content_line):
            decoded_text = decode_pdf_array_text(array_match.group(1), unicode_map)
            if decoded_text.strip():
                text_items.append({"x": current_text_x, "y": current_text_y, "text": decoded_text})

        for literal_match in re.finditer(r"\(((?:\\.|[^\\)])*)\)\s*Tj", content_line):
            decoded_text = decode_pdf_literal_text(literal_match.group(1))
            if decoded_text.strip():
                text_items.append({"x": current_text_x, "y": current_text_y, "text": decoded_text})

        for hex_match in re.finditer(r"<([0-9A-F]+)>\s*Tj", content_line, flags=re.IGNORECASE):
            decoded_text = decode_pdf_hex_text(hex_match.group(1), unicode_map)
            if decoded_text.strip():
                text_items.append({"x": current_text_x, "y": current_text_y, "text": decoded_text})

    return text_items


def group_text_items_into_lines(text_items: list[dict[str, object]]) -> list[str]:
    """Группирует отдельные текстовые элементы в строки документа."""

    sorted_items = sorted(
        text_items,
        key=lambda item: (-float(item["y"]), float(item["x"])),
    )
    grouped_lines: list[dict[str, object]] = []

    for text_item in sorted_items:
        same_line = next(
            (line for line in grouped_lines if abs(float(line["y"]) - float(text_item["y"])) < 2.5),
            None,
        )
        if same_line is None:
            grouped_lines.append({"y": text_item["y"], "items": [text_item]})
        else:
            same_line["items"].append(text_item)

    result_lines: list[str] = []
    for grouped_line in sorted(grouped_lines, key=lambda line: -float(line["y"])):
        line_text = "".join(
            str(item["text"])
            for item in sorted(grouped_line["items"], key=lambda line_item: float(line_item["x"]))
        )
        line_text = re.sub(r"\s+", " ", line_text).strip()
        if line_text:
            result_lines.append(line_text)

    return result_lines


def decode_pdf_array_text(array_source: str, unicode_map: dict[str, str]) -> str:
    """Декодирует массив TJ с hex- и literal-токенами."""

    token_pattern = re.compile(r"<([0-9A-F]+)>|\(((?:\\.|[^\\)])*)\)|(-?\d+(?:\.\d+)?)", flags=re.IGNORECASE)
    decoded_fragments: list[str] = []

    for token_match in token_pattern.finditer(array_source):
        if token_match.group(1):
            decoded_fragments.append(decode_pdf_hex_text(token_match.group(1), unicode_map))
            continue

        if token_match.group(2):
            decoded_fragments.append(decode_pdf_literal_text(token_match.group(2)))

    return "".join(decoded_fragments)


def decode_pdf_literal_text(literal_text: str) -> str:
    """Декодирует literal-строку PDF."""

    return (
        literal_text
        .replace(r"\(", "(")
        .replace(r"\)", ")")
        .replace(r"\n", "\n")
        .replace(r"\r", "\r")
        .replace(r"\t", "\t")
        .replace(r"\\", "\\")
    )


def decode_pdf_hex_text(hex_text: str, unicode_map: dict[str, str]) -> str:
    """Декодирует hex-последовательность PDF через ToUnicode-карту."""

    normalized_hex = hex_text.upper()
    glyph_chunks = re.findall(r".{1,4}", normalized_hex)
    return "".join(unicode_map.get(glyph_chunk, "") for glyph_chunk in glyph_chunks)


def source_code_to_key(source_code: int, width: int) -> str:
    """Преобразует числовой код символа в hex-ключ CMap фиксированной ширины."""

    return format(source_code, "x").upper().zfill(width)


def unicode_from_hex(unicode_hex: str) -> str:
    """Преобразует hex-код Unicode в Python-символ."""

    return chr(int(unicode_hex, 16))

File: tools/cv-importer/native/src/pdf_text_extractor.cpp
#include "portfolio_cv_importer/pdf_text_extractor.hpp"

#include "portfolio_cv_importer/pdf_object_reader.hpp"
#include "portfolio_cv_importer/pdf_stream_decoder.hpp"

#include <algorithm>
#include <cmath>
#include <fstream>
#include <optional>
#include <regex>
#include <sstream>
#include <stdexcept>
#include <unordered_map>

namespace portfolio_cv_importer {

namespace {

struct TextItem {
  double x;
  double y;
  std::string text;
};

using UnicodeMap = std::unordered_map<std::string, std::string>;
using FontUnicodeMaps = std::unordered_map<int, UnicodeMap>;
using PageFontReferences = std::unordered_map<std::string, int>;

std::string Utf8FromCodePoint(int code_point) {
  std::string result;
  if (code_point <= 0x7F) {
    result.push_back(static_cast<char>(code_point));
    return result;
  }

  if (code_point <= 0x7FF) {
    result.push_back(static_cast<char>(0xC0 | ((code_point >> 6) & 0x1F)));
    result.push_back(static_cast<char>(0x80 | (code_point & 0x3F)));
    return result;
  }

  if (code_point <= 0xFFFF) {
    result.push_back(static_cast<char>(0xE0 | ((code_point >> 12) & 0x0F)));
    result.push_back(static_cast<char>(0x80 | ((code_point >> 6) & 0x3F)));
    result.push_back(static_cast<char>(0x80 | (code_point & 0x3F)));
    return result;
  }

  result.push_back(static_cast<char>(0xF0 | ((code_point >> 18) & 0x07)));
  result.push_back(static_cast<char>(0x80 | ((code_point >> 12) & 0x3F)));
  result.push_back(static_cast<char>(0x80 | ((code_point >> 6) & 0x3F)));
  result.push_back(static_cast<char>(0x80 | (code_point & 0x3F)));
  return result;
}

std::string UnicodeFromHex(const std::string& unicode_hex) {
  return Utf8FromCodePoint(std::stoi(unicode_hex, nullptr, 16));
}

std::string SourceCodeToKey(int source_code, std::size_t width) {
  std::ostringstream output_stream;
  output_stream << std::uppercase << std::hex << source_code;
  std::string key = output_stream.str();
  if (key.size() < width) {
    key.insert(0, width - key.size(), '0');
  }
  return key;
}

std::string NormalizeExtractedLine(const std::string& raw_line) {
  std::string normalized_line = raw_line;
  const std::pair<const char*, const char*> replacements[] = {
    {"Языкипрограммированияифреймворки", "Языки программирования и фреймворки"},
    {"Управлениепроектами", "Управление проектами"},
    {"Личныекачества", "Личные качества"},
    {"Сертификатыирекомендации", "Сертификаты и рекомендации"},
    {"УЧЕБНЫЕПРОЕКТЫ", "УЧЕБНЫЕ ПРОЕКТЫ"},
    {"Бауманскаяинженернаяшкола", "Бауманская инженерная школа"},
    {"Московскийгосударственныйтехническийуниверситет", "Московский государственный технический университет"},
    {"Производственнаяпрактика", "Производственная практика"},
    {"инженернойшколе", "инженерной школе"},
    {"поднагрузкой", "под нагрузкой"},
  };

  for (const auto& [source_fragment, target_fragment] : replacements) {
    const std::string source_string(source_fragment);
    const std::string target_string(target_fragment);
    std::size_t search_position = 0;
    while ((search_position = normalized_line.find(source_string, search_position)) != std::string::npos) {
      normalized_line.replace(search_position, source_string.size(), target_string);
      search_position += target_string.size();
    }
  }

  normalized_line = std::regex_replace(normalized_line, std::regex("\\s+"), " ");
  if (!normalized_line.empty() && normalized_line.front() == ' ') {
    normalized_line.erase(normalized_line.begin());
  }
  if (!normalized_line.empty() && normalized_line.back() == ' ') {
    normalized_line.pop_back();
  }
  return normalized_line;
}

UnicodeMap ParseToUnicodeMap(const std::string& unicode_stream) {
  UnicodeMap unicode_map;
  const std::regex block_pattern(R"((?:\d+)\s+begin(?:bfchar|bfrange)[\s\S]*?end(?:bfchar|bfrange))");
  const std::regex char_pattern(R"(^<([0-9A-F]+)>\s+<([0-9A-F]+)>$)", std::regex::icase);
  const std::regex range_pattern(R"(^<([0-9A-F]+)>\s+<([0-9A-F]+)>\s+<([0-9A-F]+)>$)", std::regex::icase);
  const std::regex array_pattern(R"(^<([0-9A-F]+)>\s+<([0-9A-F]+)>\s+\[(.+)\]$)", std::regex::icase);
  const std::regex array_target_pattern(R"(<([0-9A-F]+)>)", std::regex::icase);

  for (std::sregex_iterator block_iterator(unicode_stream.begin(), unicode_stream.end(), block_pattern), block_end;
       block_iterator != block_end;
       ++block_iterator) {
    std::istringstream block_stream((*block_iterator).str());
    std::string block_line;
    while (std::getline(block_stream, block_line)) {
      if (block_line.empty()) {
        continue;
      }

      std::smatch line_match;
      if (std::regex_match(block_line, line_match, char_pattern)) {
        unicode_map[line_match[1].str()] = UnicodeFromHex(line_match[2].str());
        continue;
      }

      if (std::regex_match(block_line, line_match, range_pattern)) {
        const int range_start = std::stoi(line_match[1].str(), nullptr, 16);
        const int range_end = std::stoi(line_match[2].str(), nullptr, 16);
        int target_code = std::stoi(line_match[3].str(), nullptr, 16);

        for (int source_code = range_start; source_code <= range_end; ++source_code, ++target_code) {
          unicode_map[SourceCodeToKey(source_code, line_match[1].str().size())] = UnicodeFromHex(
            SourceCodeToKey(target_code, 1)
          );
        }
        continue;
      }

      if (!std::regex_match(block_line, line_match, array_pattern)) {
        continue;
      }

      const int range_start = std::stoi(line_match[1].str(), nullptr, 16);
      int target_index = 0;
      for (std::sregex_iterator target_iterator(line_match[3].first, line_match[3].second, array_target_pattern), target_end;
           target_iterator != target_end;
           ++target_iterator, ++target_index) {
        unicode_map[SourceCodeToKey(range_start + target_index, line_match[1].str().size())] = UnicodeFromHex(
          (*target_iterator)[1].str()
        );
      }
    }
  }

  return unicode_map;
}

FontUnicodeMaps BuildFontToUnicodeMaps(const PdfObjectMap& pdf_objects) {
  FontUnicodeMaps font_maps;
  const std::regex to_unicode_pattern(R"(\/ToUnicode\s+(\d+)\s+0\s+R)");

  for (const auto& [object_id, body_text] : pdf_objects) {
    std::smatch to_unicode_match;
    if (!std::regex_search(body_text, to_unicode_match, to_unicode_pattern)) {
      continue;
    }

    const auto unicode_object_iterator = pdf_objects.find(std::stoi(to_unicode_match[1].str()));
    if (unicode_object_iterator == pdf_objects.end()) {
      continue;
    }

    const std::optional<std::string> unicode_stream = InflateObjectStream(unicode_object_iterator->second);
    if (!unicode_stream.has_value()) {
      continue;
    }

    font_maps[object_id] = ParseToUnicodeMap(unicode_stream.value());
  }

  return font_maps;
}

PageFontReferences ParsePageFontReferences(const std::string& body_text) {
  PageFontReferences font_references;
  const std::regex font_block_pattern(R"(\/Font\s*<<([\s\S]*?)>>)");
  const std::regex font_reference_pattern(R"(\/(F\d+)\s+(\d+)\s+0\s+R)");
  std::smatch font_block_match;

  if (!std::regex_search(body_text, font_block_match, font_block_pattern)) {
    return font_references;
  }

  for (std::sregex_iterator font_iterator(font_block_match[1].first, font_block_match[1].second, font_reference_pattern), font_end;
       font_iterator != font_end;
       ++font_iterator) {
    font_references[(*font_iterator)[1].str()] = std::stoi((*font_iterator)[2].str());
  }

  return font_references;
}

std::string DecodePdfLiteralText(const std::string& literal_text) {
  std::string decoded_text = literal_text;
  const std::pair<const char*, const char*> replacements[] = {
    {R"(\()", "("},
    {R"(\))", ")"},
    {R"(\n)", "\n"},
    {R"(\r)", "\r"},
    {R"(\t)", "\t"},
    {R"(\\)", R"(\)"},
  };

  for (const auto& [source_fragment, target_fragment] : replacements) {
    const std::string source_string(source_fragment);
    const std::string target_string(target_fragment);
    std::size_t search_position = 0;
    while ((search_position = decoded_text.find(source_string, search_position)) != std::string::npos) {
      decoded_text.replace(search_position, source_string.size(), target_string);
      search_position += target_string.size();
    }
  }

  return decoded_text;
}

std::string DecodePdfHexText(const std::string& hex_text, const UnicodeMap& unicode_map) {
  std::string decoded_text;
  for (std::size_t index = 0; index < hex_text.size(); index += 4) {
    const std::string glyph_chunk = hex_text.substr(index, std::min<std::size_t>(4, hex_text.size() - index));
    const auto glyph_iterator = unicode_map.find(glyph_chunk);
    if (glyph_iterator != unicode_map.end()) {
      decoded_text.append(glyph_iterator->second);
    }
  }
  return decoded_text;
}

std::string DecodePdfArrayText(const std::string& array_source, const UnicodeMap& unicode_map) {
  const std::regex token_pattern(R"(<([0-9A-F]+)>|\(((?:\\.|[^\\)])*)\)|(-?\d+(?:\.\d+)?))", std::regex::icase);
  std::string decoded_text;

  for (std::sregex_iterator token_iterator(array_source.begin(), array_source.end(), token_pattern), token_end;
       token_iterator != token_end;
       ++token_iterator) {
    if ((*token_iterator)[1].matched) {
      decoded_text.append(DecodePdfHexText((*token_iterator)[1].str(), unicode_map));
      continue;
    }

    if ((*token_iterator)[2].matched) {
      decoded_text.append(DecodePdfLiteralText((*token_iterator)[2].str()));
    }
  }

  return decoded_text;
}

std::vector<TextItem> ExtractTextItems(
  const std::string& content_stream,
  const PageFontReferences& page_font_references,
  const FontUnicodeMaps& font_to_unicode_maps
) {
  std::vector<TextItem> text_items;
  const std::regex font_pattern(R"(\/(F\d+)\s+[\d.]+\s+Tf)");
  const std::regex transform_pattern(R"(([\d.-]+)\s+[\d.-]+\s+[\d.-]+\s+[\d.-]+\s+([\d.-]+)\s+([\d.-]+)\s+Tm)");
  const std::regex array_pattern(R"(\[(.*?)\]\s*TJ)");
  const std::regex literal_pattern(R"(\(((?:\\.|[^\\)])*)\)\s*Tj)");
  const std::regex hex_pattern(R"(<([0-9A-F]+)>\s*Tj)", std::regex::icase);

  std::string current_font_name;
  double current_text_x = 0.0;
  double current_text_y = 0.0;

  std::istringstream content_stream_reader(content_stream);
  std::string content_line;
  while (std::getline(content_stream_reader, content_line)) {
    std::smatch content_match;
    if (std::regex_search(content_line, content_match, font_pattern)) {
      current_font_name = content_match[1].str();
    }

    if (std::regex_search(content_line, content_match, transform_pattern)) {
      current_text_x = std::stod(content_match[2].str());
      current_text_y = std::stod(content_match[3].str());
    }

    UnicodeMap unicode_map;
    const auto font_reference_iterator = page_font_references.find(current_font_name);
    if (font_reference_iterator != page_font_references.end()) {
      const auto unicode_map_iterator = font_to_unicode_maps.find(font_reference_iterator->second);
      if (unicode_map_iterator != font_to_unicode_maps.end()) {
        unicode_map = unicode_map_iterator->second;
      }
    }

    for (std::sregex_iterator array_iterator(content_line.begin(), content_line.end(), array_pattern), array_end;
         array_iterator != array_end;
         ++array_iterator) {
      const std::string decoded_text = DecodePdfArrayText((*array_iterator)[1].str(), unicode_map);
      if (!decoded_text.empty()) {
        text_items.push_back({current_text_x, current_text_y, decoded_text});
      }
    }

    for (std::sregex_iterator literal_iterator(content_line.begin(), content_line.end(), literal_pattern), literal_end;
         literal_iterator != literal_end;
         ++literal_iterator) {
      const std::string decoded_text = DecodePdfLiteralText((*literal_iterator)[1].str());
      if (!decoded_text.empty()) {
        text_items.push_back({current_text_x, current_text_y, decoded_text});
      }
    }

    for (std::sregex_iterator hex_iterator(content_line.begin(), content_line.end(), hex_pattern), hex_end;
         hex_iterator != hex_end;
         ++hex_iterator) {
      const std::string decoded_text = DecodePdfHexText((*hex_iterator)[1].str(), unicode_map);
      if (!decoded_text.empty()) {
        text_items.push_back({current_text_x, current_text_y, decoded_text});
      }
    }
  }

  return text_items;
}

std::vector<std::string> GroupTextItemsIntoLines(const std::vector<TextItem>& text_items) {
  std::vector<TextItem> sorted_items = text_items;
  std::sort(
    sorted_items.begin(),
    sorted_items.end(),
    [](const TextItem& left_item, const TextItem& right_item) {
      if (left_item.y != right_item.y) {
        return left_item.y > right_item.y;
      }
      return left_item.x < right_item.x;
    }
  );

  struct GroupedLine {
    double y;
    std::vector<TextItem> items;
  };

  std::vector<GroupedLine> grouped_lines;
  for (const TextItem& text_item : sorted_items) {
    auto same_line_iterator = std::find_if(
      grouped_lines.begin(),
      grouped_lines.end(),
      [&text_item](const GroupedLine& grouped_line) {
        return std::fabs(grouped_line.y - text_item.y) < 2.5;
      }
    );

    if (same_line_iterator == grouped_lines.end()) {
      grouped_lines.push_back({text_item.y, {text_item}});
      continue;
    }

    same_line_iterator->items.push_back(text_item);
  }

  std::sort(
    grouped_lines.begin(),
    grouped_lines.end(),
    [](const GroupedLine& left_line, const GroupedLine& right_line) {
      return left_line.y > right_line.y;
    }
  );

  std::vector<std::string> result_lines;
  for (GroupedLine& grouped_line : grouped_lines) {
    std::sort(
      grouped_line.items.begin(),
      grouped_line.items.end(),
      [](const TextItem& left_item, const TextItem& right_item) {
        return left_item.x < right_item.x;
      }
    );

    std::string line_text;
    for (const TextItem& text_item : grouped_line.items) {
      line_text.append(text_item.text);
    }

    line_text = NormalizeExtractedLine(line_text);
    if (!line_text.empty()) {
      result_lines.push_back(line_text);
    }
  }

  return result_lines;
}

std::vector<unsigned char> ReadBinaryFile(const std::filesystem::path& pdf_path) {
  std::ifstream file_stream(pdf_path, std::ios::binary);
  if (!file_stream.is_open()) {
    throw std::runtime_error("Could not open PDF file for reading.");
  }

  return std::vector<unsigned char>(
    std::istreambuf_iterator<char>(file_stream),
    std::istreambuf_iterator<char>()
  );
}

}  // namespace

std::vector<std::string> ExtractLinesFromPdfFile(const std::filesystem::path& pdf_path) {
  return ExtractLinesFromPdfBytes(ReadBinaryFile(pdf_path));
}

std::vector<std::string> ExtractLinesFromPdfBytes(const std::vector<unsigned char>& pdf_bytes) {
  const std::string pdf_text(pdf_bytes.begin(), pdf_bytes.end());
  const PdfObjectMap pdf_objects = ParsePdfObjects(pdf_text);
  const FontUnicodeMaps font_to_unicode_maps = BuildFontToUnicodeMaps(pdf_objects);
  std::vector<std::string> page_lines;

  for (const auto& [object_id, body_text] : pdf_objects) {
    (void)object_id;
    if (body_text.find("/Type/Page") == std::string::npos) {
      continue;
    }

    const std::regex content_reference_pattern(R"(\/Contents\s+(\d+)\s+0\s+R)");
    std::smatch content_reference_match;
    if (!std::regex_search(body_text, content_reference_match, content_reference_pattern)) {
      continue;
    }

    const PageFontReferences page_font_references = ParsePageFontReferences(body_text);
    const auto content_object_iterator = pdf_objects.find(std::stoi(content_reference_match[1].str()));
    if (content_object_iterator == pdf_objects.end()) {
      continue;
    }

    const std::optional<std::string> content_stream = InflateObjectStream(content_object_iterator->second);
    if (!content_stream.has_value()) {
      continue;
    }

    const std::vector<TextItem> text_items = ExtractTextItems(
      content_stream.value(),
      page_font_references,
      font_to_unicode_maps
    );
    const std::vector<std::string> grouped_lines = GroupTextItemsIntoLines(text_items);
    page_lines.insert(page_lines.end(), grouped_lines.begin(), grouped_lines.end());
  }

  return page_lines;
}

}  // namespace portfolio_cv_importer
